src/bvals/cc/bvals_cc_in_one.cpp

```diff
@@ -0,0 +1,553 @@
+                  });
+            });
+      });
+
+  Kokkos::fence();
+  Kokkos::Profiling::pushRegion("Set complete and/or start sending via MPI");
+  for (int block = 0; block < md->NumBlocks(); block++) {
+    auto &rc = md->GetBlockData(block);
+    auto pmb = rc->GetBlockPointer();
+
+    int mylevel = pmb->loc.level;
+    for (int n = 0; n < pmb->pbval->nneighbor; n++) {
+      parthenon::NeighborBlock &nb = pmb->pbval->neighbor[n];
+      // TODO(?) currently this only works for a single "Variable" per container.
+      // Need to update the buffer sizes so that it matches the packed Variables.
+      auto *bd_var_ = rc->GetCellVariableVector()[0]->vbvar->GetPBdVar();
+      if (bd_var_->sflag[nb.bufid] == parthenon::BoundaryStatus::completed) continue;
+
+      // on the same rank the data has been directly copied to the target buffer
+      if (nb.snb.rank == parthenon::Globals::my_rank) {
+        // TODO(?) check performance of FindMeshBlock. Could be caching from call above.
+        auto target_block = pmb->pmy_mesh->FindMeshBlock(nb.snb.gid);
+        target_block->pbval->bvars[0]->GetPBdVar()->flag[nb.targetid] =
+            parthenon::BoundaryStatus::arrived;
+      } else {
+#ifdef MPI_PARALLEL
+        MPI_Start(&(bd_var_->req_send[nb.bufid]));
+#endif
+      }
+
+      bd_var_->sflag[nb.bufid] = parthenon::BoundaryStatus::completed;
+    }
+  }
+  Kokkos::Profiling::popRegion(); // Set complete and/or start sending via MPI
+
+  Kokkos::Profiling::popRegion(); // Task_SendBoundaryBuffers_MeshData
+  // TODO(?) reintroduce sparse logic (or merge with above)
+  return TaskStatus::complete;
+}
+
+TaskStatus ReceiveBoundaryBuffers(std::shared_ptr<MeshData<Real>> &md) {
+  Kokkos::Profiling::pushRegion("Task_ReceiveBoundaryBuffers_MeshData");
+  bool ret = true;
+  for (int i = 0; i < md->NumBlocks(); i++) {
+    auto &rc = md->GetBlockData(i);
+    auto task_status = rc->ReceiveBoundaryBuffers();
+    if (task_status == TaskStatus::incomplete) {
+      ret = false;
+    }
+  }
+
+  // TODO(?) reintroduce sparse logic (or merge with above)
+  Kokkos::Profiling::popRegion(); // Task_ReceiveBoundaryBuffers_MeshData
+  if (ret) return TaskStatus::complete;
+  return TaskStatus::incomplete;
+}
+
+// set boundaries from buffers with MeshBlockPack support
+// TODO(pgrete) should probaly be moved to the bvals or interface folders
+TaskStatus SetBoundaries(std::shared_ptr<MeshData<Real>> &md) {
+  Kokkos::Profiling::pushRegion("Task_SetBoundaries_MeshData");
+
+  IndexDomain interior = IndexDomain::interior;
+
+  // first calculate the number of active buffers
+  int buffers_used = 0;
+  for (int block = 0; block < md->NumBlocks(); block++) {
+    auto &rc = md->GetBlockData(block);
+    auto pmb = rc->GetBlockPointer();
+    for (int n = 0; n < pmb->pbval->nneighbor; n++) {
+      buffers_used += 1;
+    }
+  }
+
+  auto boundary_info = md->GetSetBuffers();
+  if (!boundary_info.is_allocated()) {
```

Comment:
The "set boundary info" parts of this functionality, for both this function and the one above, appear to me to be different functionality from the rest of this task. Could they be split out? That would make this much easier to follow I think.

Reply:
They are kind of related, but could be separated, see my longer reply above.